{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/bio/source/notebooks'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'Cell Passage.ipynb'   PyVCF.ipynb   Untitled.ipynb\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data\n"
     ]
    }
   ],
   "source": [
    "cd /data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Deepak_HEX52.variant_depth.example.xlsx\r\n",
      "'Fwd _exome_.zip'\r\n",
      " \u001b[0m\u001b[01;34mIllumina\u001b[0m/\r\n",
      " data.zip\r\n",
      " gene_mutation_sample_count.xlsx\r\n",
      " \u001b[01;34mindisulam\u001b[0m/\r\n",
      "\u001b[01;34m'new exomes'\u001b[0m/\r\n",
      " read_count.xlsx\r\n",
      "\u001b[01;34m'second compound'\u001b[0m/\r\n",
      " variant_allele_frequency.annotation.xlsx\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/new exomes\n"
     ]
    }
   ],
   "source": [
    "cd 'new exomes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SW246528.gene_mutation_sample_count.xlsx\r\n",
      "SW246528.variant_allele_frequency.annotation.xlsx\r\n",
      "SW252049.gene_mutation_sample_count.xlsx\r\n",
      "SW252049.variant_allele_frequency.annotation.xlsx\r\n",
      "SW271190.gene_mutation_sample_count.xlsx\r\n",
      "SW271190.variant_allele_frequency.annotation.xlsx\r\n",
      "SW388710.gene_mutation_sample_count.xlsx\r\n",
      "SW388710.variant_allele_frequency.annotation.xlsx\r\n",
      "read_variant_count.xlsx\r\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference base</th>\n",
       "      <th>Alternate base</th>\n",
       "      <th>278 genotype</th>\n",
       "      <th>278 depth</th>\n",
       "      <th>278 reference base depth</th>\n",
       "      <th>278 alternate base depth</th>\n",
       "      <th>278 variant allele frequency</th>\n",
       "      <th>314 genotype</th>\n",
       "      <th>...</th>\n",
       "      <th>COSMIC Coding mutation ID</th>\n",
       "      <th>COSMIC Coding occurrence</th>\n",
       "      <th>COSMIC Coding primary site</th>\n",
       "      <th>ClinVar disease</th>\n",
       "      <th>ClinVar significance</th>\n",
       "      <th>gnomAD allele frequency</th>\n",
       "      <th>gnomAD count of homozygous individuals</th>\n",
       "      <th>CADD raw score</th>\n",
       "      <th>CADD PHRED-scaled score</th>\n",
       "      <th>Tandem repeat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2629379</td>\n",
       "      <td>CCAG</td>\n",
       "      <td>C</td>\n",
       "      <td>0/0</td>\n",
       "      <td>47</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0/1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.683769</td>\n",
       "      <td>16.770</td>\n",
       "      <td>(CAG)7&gt;6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6469122</td>\n",
       "      <td>TTCC</td>\n",
       "      <td>T</td>\n",
       "      <td>0/0</td>\n",
       "      <td>62</td>\n",
       "      <td>50</td>\n",
       "      <td>6</td>\n",
       "      <td>0.097</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Distal_spinal_muscular_atrophy|not_specified</td>\n",
       "      <td>Benign/Likely_benign</td>\n",
       "      <td>0.089700</td>\n",
       "      <td>684.0</td>\n",
       "      <td>1.303892</td>\n",
       "      <td>14.530</td>\n",
       "      <td>(TCC)8&gt;7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10948808</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0/0</td>\n",
       "      <td>92</td>\n",
       "      <td>91</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.537956</td>\n",
       "      <td>0.039</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11122013</td>\n",
       "      <td>ATCT</td>\n",
       "      <td>A</td>\n",
       "      <td>0/0</td>\n",
       "      <td>73</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0/1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.584805</td>\n",
       "      <td>22.200</td>\n",
       "      <td>(TCT)4&gt;3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11958532</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0/0</td>\n",
       "      <td>122</td>\n",
       "      <td>122</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ehlers-Danlos_syndrome,_hydroxylysine-deficient</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.503887</td>\n",
       "      <td>7.363</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>641</th>\n",
       "      <td>chrX</td>\n",
       "      <td>141906056</td>\n",
       "      <td>CCTGAGAGAACTCAGAGTACTTTTGAGGGTTTTGCCCAGTCTCCTC...</td>\n",
       "      <td>C</td>\n",
       "      <td>0/1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0/1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.012100</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.220570</td>\n",
       "      <td>4.328</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642</th>\n",
       "      <td>chrX</td>\n",
       "      <td>147930143</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0/0</td>\n",
       "      <td>59</td>\n",
       "      <td>59</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.868358</td>\n",
       "      <td>22.900</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>chrX</td>\n",
       "      <td>150987884</td>\n",
       "      <td>AGAGGAG</td>\n",
       "      <td>[A],AGAG</td>\n",
       "      <td>2/2</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2/2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.062758</td>\n",
       "      <td>19.130</td>\n",
       "      <td>(GAG)7&gt;5,(GAG)7&gt;6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154097642</td>\n",
       "      <td>CGCG</td>\n",
       "      <td>C</td>\n",
       "      <td>0/0</td>\n",
       "      <td>39</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>History_of_neurodevelopmental_disorder</td>\n",
       "      <td>Benign</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.384953</td>\n",
       "      <td>21.300</td>\n",
       "      <td>(GCG)6&gt;5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154776813</td>\n",
       "      <td>CAAG</td>\n",
       "      <td>C</td>\n",
       "      <td>0/1</td>\n",
       "      <td>36</td>\n",
       "      <td>30</td>\n",
       "      <td>5</td>\n",
       "      <td>0.139</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dyskeratosis_congenita|Dyskeratosis_congenita_...</td>\n",
       "      <td>Conflicting_interpretations_of_pathogenicity</td>\n",
       "      <td>0.000958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.891580</td>\n",
       "      <td>10.850</td>\n",
       "      <td>(AAG)7&gt;6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>646 rows Ã— 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome   Position                                     Reference base  \\\n",
       "0         chr1    2629379                                               CCAG   \n",
       "1         chr1    6469122                                               TTCC   \n",
       "2         chr1   10948808                                                  C   \n",
       "3         chr1   11122013                                               ATCT   \n",
       "4         chr1   11958532                                                  C   \n",
       "..         ...        ...                                                ...   \n",
       "641       chrX  141906056  CCTGAGAGAACTCAGAGTACTTTTGAGGGTTTTGCCCAGTCTCCTC...   \n",
       "642       chrX  147930143                                                  A   \n",
       "643       chrX  150987884                                            AGAGGAG   \n",
       "644       chrX  154097642                                               CGCG   \n",
       "645       chrX  154776813                                               CAAG   \n",
       "\n",
       "    Alternate base 278 genotype  278 depth  278 reference base depth  \\\n",
       "0                C          0/0         47                        22   \n",
       "1                T          0/0         62                        50   \n",
       "2                T          0/0         92                        91   \n",
       "3                A          0/0         73                        71   \n",
       "4                T          0/0        122                       122   \n",
       "..             ...          ...        ...                       ...   \n",
       "641              C          0/1          0                         0   \n",
       "642              G          0/0         59                        59   \n",
       "643       [A],AGAG          2/2         11                         0   \n",
       "644              C          0/0         39                        26   \n",
       "645              C          0/1         36                        30   \n",
       "\n",
       "     278 alternate base depth  278 variant allele frequency 314 genotype  ...  \\\n",
       "0                           1                         0.021          0/1  ...   \n",
       "1                           6                         0.097          0/0  ...   \n",
       "2                           0                         0.000          0/0  ...   \n",
       "3                           0                         0.000          0/1  ...   \n",
       "4                           0                         0.000          0/0  ...   \n",
       "..                        ...                           ...          ...  ...   \n",
       "641                         0                           NaN          0/1  ...   \n",
       "642                         0                         0.000          0/0  ...   \n",
       "643                         0                         0.000          2/2  ...   \n",
       "644                         0                         0.000          0/0  ...   \n",
       "645                         5                         0.139          0/0  ...   \n",
       "\n",
       "     COSMIC Coding mutation ID  COSMIC Coding occurrence  \\\n",
       "0                          NaN                       NaN   \n",
       "1                          NaN                       NaN   \n",
       "2                          NaN                       NaN   \n",
       "3                          NaN                       NaN   \n",
       "4                          NaN                       NaN   \n",
       "..                         ...                       ...   \n",
       "641                        NaN                       NaN   \n",
       "642                        NaN                       NaN   \n",
       "643                        NaN                       NaN   \n",
       "644                        NaN                       NaN   \n",
       "645                        NaN                       NaN   \n",
       "\n",
       "     COSMIC Coding primary site  \\\n",
       "0                           NaN   \n",
       "1                           NaN   \n",
       "2                           NaN   \n",
       "3                           NaN   \n",
       "4                           NaN   \n",
       "..                          ...   \n",
       "641                         NaN   \n",
       "642                         NaN   \n",
       "643                         NaN   \n",
       "644                         NaN   \n",
       "645                         NaN   \n",
       "\n",
       "                                       ClinVar disease  \\\n",
       "0                                                  NaN   \n",
       "1         Distal_spinal_muscular_atrophy|not_specified   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4      Ehlers-Danlos_syndrome,_hydroxylysine-deficient   \n",
       "..                                                 ...   \n",
       "641                                                NaN   \n",
       "642                                                NaN   \n",
       "643                                                NaN   \n",
       "644             History_of_neurodevelopmental_disorder   \n",
       "645  Dyskeratosis_congenita|Dyskeratosis_congenita_...   \n",
       "\n",
       "                             ClinVar significance  gnomAD allele frequency  \\\n",
       "0                                             NaN                 0.000329   \n",
       "1                            Benign/Likely_benign                 0.089700   \n",
       "2                                             NaN                      NaN   \n",
       "3                                             NaN                 0.000021   \n",
       "4                          Uncertain_significance                 0.000014   \n",
       "..                                            ...                      ...   \n",
       "641                                           NaN                 0.012100   \n",
       "642                                           NaN                      NaN   \n",
       "643                                           NaN                 0.000220   \n",
       "644                                        Benign                 0.000010   \n",
       "645  Conflicting_interpretations_of_pathogenicity                 0.000958   \n",
       "\n",
       "     gnomAD count of homozygous individuals  CADD raw score  \\\n",
       "0                                       0.0        1.683769   \n",
       "1                                     684.0        1.303892   \n",
       "2                                       NaN       -0.537956   \n",
       "3                                       0.0        2.584805   \n",
       "4                                       0.0        0.503887   \n",
       "..                                      ...             ...   \n",
       "641                                     2.0        0.220570   \n",
       "642                                     NaN        2.868358   \n",
       "643                                     0.0        2.062758   \n",
       "644                                     0.0        2.384953   \n",
       "645                                     0.0        0.891580   \n",
       "\n",
       "     CADD PHRED-scaled score      Tandem repeat  \n",
       "0                     16.770           (CAG)7>6  \n",
       "1                     14.530           (TCC)8>7  \n",
       "2                      0.039                NaN  \n",
       "3                     22.200           (TCT)4>3  \n",
       "4                      7.363                NaN  \n",
       "..                       ...                ...  \n",
       "641                    4.328                NaN  \n",
       "642                   22.900                NaN  \n",
       "643                   19.130  (GAG)7>5,(GAG)7>6  \n",
       "644                   21.300           (GCG)6>5  \n",
       "645                   10.850           (AAG)7>6  \n",
       "\n",
       "[646 rows x 64 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('SW246528.variant_allele_frequency.annotation.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Package              Version\n",
      "-------------------- ----------------------------\n",
      "absl-py              0.9.0\n",
      "argon2-cffi          20.1.0\n",
      "astor                0.8.1\n",
      "attrs                20.1.0\n",
      "backcall             0.2.0\n",
      "biopython            1.76\n",
      "bleach               3.1.5\n",
      "certifi              2020.6.20\n",
      "cffi                 1.14.2\n",
      "cycler               0.10.0\n",
      "decorator            4.4.2\n",
      "defusedxml           0.6.0\n",
      "entrypoints          0.3\n",
      "gast                 0.2.2\n",
      "gatkpythonpackages   0.1\n",
      "google-pasta         0.2.0\n",
      "grpcio               1.30.0\n",
      "h5py                 2.10.0\n",
      "importlib-metadata   1.7.0\n",
      "ipykernel            5.3.4\n",
      "ipython              7.16.1\n",
      "ipython-genutils     0.2.0\n",
      "ipywidgets           7.5.1\n",
      "jedi                 0.17.2\n",
      "Jinja2               2.11.2\n",
      "joblib               0.16.0\n",
      "jsonschema           3.2.0\n",
      "jupyter              1.0.0\n",
      "jupyter-client       6.1.7\n",
      "jupyter-console      6.1.0\n",
      "jupyter-core         4.6.3\n",
      "Keras                2.2.4\n",
      "Keras-Applications   1.0.8\n",
      "Keras-Preprocessing  1.1.0\n",
      "kiwisolver           1.2.0\n",
      "Mako                 1.1.0\n",
      "Markdown             3.2.2\n",
      "MarkupSafe           1.1.1\n",
      "matplotlib           3.2.1\n",
      "mistune              0.8.4\n",
      "mkl-service          2.3.0\n",
      "mpmath               1.1.0\n",
      "nbconvert            5.6.1\n",
      "nbformat             5.0.7\n",
      "nose                 1.3.7\n",
      "notebook             6.1.3\n",
      "numpy                1.17.5\n",
      "opt-einsum           0+untagged.56.g2664021.dirty\n",
      "packaging            20.4\n",
      "pandas               1.0.3\n",
      "pandocfilters        1.4.2\n",
      "parso                0.7.1\n",
      "patsy                0.5.1\n",
      "pexpect              4.8.0\n",
      "pickleshare          0.7.5\n",
      "pip                  20.2.2\n",
      "prometheus-client    0.8.0\n",
      "prompt-toolkit       3.0.6\n",
      "protobuf             3.12.3\n",
      "ptyprocess           0.6.0\n",
      "pycparser            2.20\n",
      "Pygments             2.6.1\n",
      "pygpu                0.7.6\n",
      "pymc3                3.1\n",
      "pyparsing            2.4.7\n",
      "PyQt5                5.12.3\n",
      "PyQt5-sip            4.19.18\n",
      "PyQtChart            5.12\n",
      "PyQtWebEngine        5.12.1\n",
      "pyrsistent           0.16.0\n",
      "pysam                0.15.3\n",
      "python-dateutil      2.8.1\n",
      "pytz                 2020.1\n",
      "PyVCF                0.6.8\n",
      "PyYAML               5.3.1\n",
      "pyzmq                19.0.2\n",
      "qtconsole            4.7.6\n",
      "QtPy                 1.9.0\n",
      "scikit-learn         0.22.2.post1\n",
      "scipy                1.0.0\n",
      "Send2Trash           1.5.0\n",
      "setuptools           49.2.0.post20200712\n",
      "six                  1.15.0\n",
      "sympy                1.6.2\n",
      "tensorboard          1.15.0\n",
      "tensorflow           1.15.0\n",
      "tensorflow-estimator 1.15.1\n",
      "termcolor            1.1.0\n",
      "terminado            0.8.3\n",
      "testpath             0.4.4\n",
      "Theano               1.0.4\n",
      "tornado              6.0.4\n",
      "tqdm                 4.48.0\n",
      "traitlets            4.3.3\n",
      "wcwidth              0.2.5\n",
      "webencodings         0.5.1\n",
      "Werkzeug             0.16.1\n",
      "wheel                0.34.2\n",
      "widgetsnbextension   3.5.1\n",
      "wrapt                1.12.1\n",
      "zipp                 3.1.0\n",
      "\u001b[33mWARNING: You are using pip version 20.2.2; however, version 20.2.4 is available.\n",
      "You should consider upgrading via the '/opt/miniconda/envs/gatk/bin/python3.6 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting xlrd\n",
      "  Downloading xlrd-1.2.0-py2.py3-none-any.whl (103 kB)\n",
      "\u001b[K     |################################| 103 kB 3.4 MB/s eta 0:00:01\n",
      "\u001b[?25hInstalling collected packages: xlrd\n",
      "Successfully installed xlrd-1.2.0\n",
      "\u001b[33mWARNING: You are using pip version 20.2.2; however, version 20.2.4 is available.\n",
      "You should consider upgrading via the '/opt/miniconda/envs/gatk/bin/python3.6 -m pip install --upgrade pip' command.\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install xlrd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "SW246528_variants = pd.read_excel('SW246528.variant_allele_frequency.annotation.xlsx', sheet_name='Sheet1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
