{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/bio/source/notebooks'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "vcfDir='/data/nijhawanlab/indisulam/'\n",
    "vcfFile='D0_ACTB2.annotated.vcf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=chr1, POS=14248, REF=T, ALT=[G])\n",
      "{'AC': [2], 'AF': [1.0], 'AN': 2, 'DP': 2, 'ExcessHet': 3.0103, 'FS': 0.0, 'MLEAC': [1], 'MLEAF': [0.5], 'MQ': 28.16, 'QD': 27.14, 'SOR': 2.303, 'Annomen.geneName': ['DDX11L1'], 'Annomen.region': ['non_coding_exon'], 'Annomen.strand': ['+'], 'Annomen.spliceDistance': ['-162'], 'Annomen.transcriptId': ['NR_046018.2'], 'Annomen.proteinId': [''], 'Annomen.nucleotideVariationNomenclature': ['r.1491T>G'], 'snp151.name': ['rs1258110102'], 'snp151.subset': [''], 'snp151.alleleN': [''], 'snp151.alleleFreq': [''], 'CADD.RawScore': ['0.209183'], 'CADD.PHRED': ['4.181']}\n",
      "Call(sample=D0_ACTB2, CallData(GT=1/1, AD=[0, 2], DP=2, GQ=6, PL=[66, 6, 0]))\n",
      "[Call(sample=D0_ACTB2, CallData(GT=1/1, AD=[0, 2], DP=2, GQ=6, PL=[66, 6, 0]))]\n",
      "\n",
      "Record(CHROM=chr1, POS=14542, REF=A, ALT=[G])\n",
      "{'AC': [1], 'AF': [0.5], 'AN': 2, 'BaseQRankSum': 3.77, 'DP': 22, 'ExcessHet': 3.0103, 'FS': 0.0, 'MLEAC': [1], 'MLEAF': [0.5], 'MQ': 26.98, 'MQRankSum': -1.353, 'QD': 1.8, 'ReadPosRankSum': 1.7, 'SOR': 0.067, 'Annomen.geneName': ['WASH7P'], 'Annomen.region': ['non_coding_exon'], 'Annomen.strand': ['-'], 'Annomen.spliceDistance': ['-181'], 'Annomen.transcriptId': ['NR_024540.1'], 'Annomen.proteinId': [''], 'Annomen.nucleotideVariationNomenclature': ['r.1589T>C'], 'snp151.name': ['rs1045833'], 'snp151.subset': [''], 'snp151.alleleN': [''], 'snp151.alleleFreq': [''], 'CADD.RawScore': ['0.116314'], 'CADD.PHRED': ['2.989']}\n",
      "Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[18, 4], DP=22, GQ=47, PL=[47, 0, 474]))\n",
      "[Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[18, 4], DP=22, GQ=47, PL=[47, 0, 474]))]\n",
      "\n",
      "Record(CHROM=chr1, POS=14574, REF=A, ALT=[G])\n",
      "{'AC': [1], 'AF': [0.5], 'AN': 2, 'BaseQRankSum': -3.188, 'DP': 34, 'ExcessHet': 3.0103, 'FS': 0.0, 'MLEAC': [1], 'MLEAF': [0.5], 'MQ': 30.14, 'MQRankSum': -3.0, 'QD': 3.43, 'ReadPosRankSum': -1.341, 'SOR': 0.392, 'Annomen.geneName': ['WASH7P'], 'Annomen.region': ['non_coding_exon'], 'Annomen.strand': ['-'], 'Annomen.spliceDistance': ['-213'], 'Annomen.transcriptId': ['NR_024540.1'], 'Annomen.proteinId': [''], 'Annomen.nucleotideVariationNomenclature': ['r.1557T>C'], 'snp151.name': ['rs28503599'], 'snp151.subset': [''], 'snp151.alleleN': ['3'], 'snp151.alleleFreq': ['0.375'], 'CADD.RawScore': ['0.347491'], 'CADD.PHRED': ['5.831']}\n",
      "Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[26, 8], DP=34, GQ=99, PL=[124, 0, 694]))\n",
      "[Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[26, 8], DP=34, GQ=99, PL=[124, 0, 694]))]\n",
      "\n",
      "Record(CHROM=chr1, POS=14599, REF=T, ALT=[A])\n",
      "{'AC': [1], 'AF': [0.5], 'AN': 2, 'BaseQRankSum': -3.054, 'DP': 43, 'ExcessHet': 3.0103, 'FS': 2.411, 'MLEAC': [1], 'MLEAF': [0.5], 'MQ': 32.45, 'MQRankSum': -3.375, 'QD': 5.06, 'ReadPosRankSum': -0.797, 'SOR': 1.688, 'Annomen.geneName': ['WASH7P'], 'Annomen.region': ['non_coding_exon'], 'Annomen.strand': ['-'], 'Annomen.spliceDistance': ['231'], 'Annomen.transcriptId': ['NR_024540.1'], 'Annomen.proteinId': [''], 'Annomen.nucleotideVariationNomenclature': ['r.1532A>T'], 'snp151.name': ['rs707680'], 'snp151.subset': [''], 'snp151.alleleN': [''], 'snp151.alleleFreq': [''], 'CADD.RawScore': ['0.269088'], 'CADD.PHRED': ['4.933']}\n",
      "Call(sample=D0_ACTB2, CallData(GT=0|1, AD=[35, 8], DP=43, GQ=99, PGT=0|1, PID=14599_T_A, PL=[225, 0, 1530], PS=14599))\n",
      "[Call(sample=D0_ACTB2, CallData(GT=0|1, AD=[35, 8], DP=43, GQ=99, PGT=0|1, PID=14599_T_A, PL=[225, 0, 1530], PS=14599))]\n",
      "\n",
      "Record(CHROM=chr1, POS=14604, REF=A, ALT=[G])\n",
      "{'AC': [1], 'AF': [0.5], 'AN': 2, 'BaseQRankSum': 4.64, 'DP': 48, 'ExcessHet': 3.0103, 'FS': 4.923, 'MLEAC': [1], 'MLEAF': [0.5], 'MQ': 32.23, 'MQRankSum': -3.411, 'QD': 4.6, 'ReadPosRankSum': -0.263, 'SOR': 2.187, 'Annomen.geneName': ['WASH7P'], 'Annomen.region': ['non_coding_exon'], 'Annomen.strand': ['-'], 'Annomen.spliceDistance': ['226'], 'Annomen.transcriptId': ['NR_024540.1'], 'Annomen.proteinId': [''], 'Annomen.nucleotideVariationNomenclature': ['r.1527T>C'], 'snp151.name': ['rs541940975'], 'snp151.subset': ['Common'], 'snp151.alleleN': ['739'], 'snp151.alleleFreq': ['0.147564'], 'CADD.RawScore': ['0.516594'], 'CADD.PHRED': ['7.477']}\n",
      "Call(sample=D0_ACTB2, CallData(GT=0|1, AD=[40, 8], DP=48, GQ=99, PGT=0|1, PID=14599_T_A, PL=[228, 0, 1718], PS=14599))\n",
      "[Call(sample=D0_ACTB2, CallData(GT=0|1, AD=[40, 8], DP=48, GQ=99, PGT=0|1, PID=14599_T_A, PL=[228, 0, 1718], PS=14599))]\n",
      "\n",
      "Record(CHROM=chr1, POS=14610, REF=T, ALT=[C])\n",
      "{'AC': [1], 'AF': [0.5], 'AN': 2, 'BaseQRankSum': 3.26, 'DP': 51, 'ExcessHet': 3.0103, 'FS': 1.763, 'MLEAC': [1], 'MLEAF': [0.5], 'MQ': 33.07, 'MQRankSum': -3.607, 'QD': 4.33, 'ReadPosRankSum': -0.853, 'SOR': 1.347, 'Annomen.geneName': ['WASH7P'], 'Annomen.region': ['non_coding_exon'], 'Annomen.strand': ['-'], 'Annomen.spliceDistance': ['220'], 'Annomen.transcriptId': ['NR_024540.1'], 'Annomen.proteinId': [''], 'Annomen.nucleotideVariationNomenclature': ['r.1521A>G'], 'snp151.name': ['rs878986575'], 'snp151.subset': [''], 'snp151.alleleN': [''], 'snp151.alleleFreq': [''], 'CADD.RawScore': ['0.957842'], 'CADD.PHRED': ['11.62']}\n",
      "Call(sample=D0_ACTB2, CallData(GT=0|1, AD=[42, 9], DP=51, GQ=99, PGT=0|1, PID=14599_T_A, PL=[228, 0, 1676], PS=14599))\n",
      "[Call(sample=D0_ACTB2, CallData(GT=0|1, AD=[42, 9], DP=51, GQ=99, PGT=0|1, PID=14599_T_A, PL=[228, 0, 1676], PS=14599))]\n",
      "\n",
      "Record(CHROM=chr1, POS=14653, REF=C, ALT=[T])\n",
      "{'AC': [1], 'AF': [0.5], 'AN': 2, 'BaseQRankSum': -1.644, 'DP': 103, 'ExcessHet': 3.0103, 'FS': 0.0, 'MLEAC': [1], 'MLEAF': [0.5], 'MQ': 36.01, 'MQRankSum': -3.183, 'QD': 6.38, 'ReadPosRankSum': -2.135, 'SOR': 0.648, 'Annomen.geneName': ['WASH7P'], 'Annomen.region': ['non_coding_exon'], 'Annomen.strand': ['-'], 'Annomen.spliceDistance': ['177'], 'Annomen.transcriptId': ['NR_024540.1'], 'Annomen.proteinId': [''], 'Annomen.nucleotideVariationNomenclature': ['r.1478G>A'], 'snp151.name': ['rs62635297'], 'snp151.subset': [''], 'snp151.alleleN': [''], 'snp151.alleleFreq': [''], 'CADD.RawScore': ['-0.373269'], 'CADD.PHRED': ['0.148']}\n",
      "Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[68, 35], DP=103, GQ=99, PL=[665, 0, 1697]))\n",
      "[Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[68, 35], DP=103, GQ=99, PL=[665, 0, 1697]))]\n",
      "\n",
      "Record(CHROM=chr1, POS=14677, REF=G, ALT=[A])\n",
      "{'AC': [1], 'AF': [0.5], 'AN': 2, 'BaseQRankSum': -2.296, 'DP': 120, 'ExcessHet': 3.0103, 'FS': 3.236, 'MLEAC': [1], 'MLEAF': [0.5], 'MQ': 36.36, 'MQRankSum': -2.1, 'QD': 2.53, 'ReadPosRankSum': -0.994, 'SOR': 1.064, 'Annomen.geneName': ['WASH7P'], 'Annomen.region': ['non_coding_exon'], 'Annomen.strand': ['-'], 'Annomen.spliceDistance': ['153'], 'Annomen.transcriptId': ['NR_024540.1'], 'Annomen.proteinId': [''], 'Annomen.nucleotideVariationNomenclature': ['r.1454C>T'], 'snp151.name': ['rs201327123'], 'snp151.subset': ['Common'], 'snp151.alleleN': ['17.978001'], 'snp151.alleleFreq': ['0.0445'], 'gnomAD.AF': ['5.00134e-02'], 'gnomAD.nhomalt': ['73'], 'CADD.RawScore': ['0.337784'], 'CADD.PHRED': ['5.725']}\n",
      "Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[96, 24], DP=120, GQ=99, PL=[311, 0, 2425]))\n",
      "[Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[96, 24], DP=120, GQ=99, PL=[311, 0, 2425]))]\n",
      "\n",
      "Record(CHROM=chr1, POS=14907, REF=A, ALT=[G])\n",
      "{'AC': [1], 'AF': [0.5], 'AN': 2, 'BaseQRankSum': 0.998, 'DP': 41, 'ExcessHet': 3.0103, 'FS': 25.375, 'MLEAC': [1], 'MLEAF': [0.5], 'MQ': 25.03, 'MQRankSum': 1.04, 'QD': 3.62, 'ReadPosRankSum': -1.418, 'SOR': 0.371, 'Annomen.geneName': ['WASH7P'], 'Annomen.region': ['non_coding_intron'], 'Annomen.strand': ['-'], 'Annomen.spliceDistance': ['63'], 'Annomen.transcriptId': ['NR_024540.1'], 'Annomen.proteinId': [''], 'snp151.name': ['rs6682375'], 'snp151.subset': [''], 'snp151.alleleN': ['1'], 'snp151.alleleFreq': ['0.5'], 'CADD.RawScore': ['0.070629'], 'CADD.PHRED': ['2.456']}\n",
      "Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[32, 9], DP=41, GQ=99, PL=[156, 0, 807]))\n",
      "[Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[32, 9], DP=41, GQ=99, PL=[156, 0, 807]))]\n",
      "\n",
      "Record(CHROM=chr1, POS=14930, REF=A, ALT=[G])\n",
      "{'AC': [1], 'AF': [0.5], 'AN': 2, 'BaseQRankSum': -3.163, 'DP': 34, 'ExcessHet': 3.0103, 'FS': 37.929, 'MLEAC': [1], 'MLEAF': [0.5], 'MQ': 25.82, 'MQRankSum': 0.833, 'QD': 3.96, 'ReadPosRankSum': -2.153, 'SOR': 2.774, 'Annomen.geneName': ['WASH7P'], 'Annomen.region': ['non_coding_intron'], 'Annomen.strand': ['-'], 'Annomen.spliceDistance': ['40'], 'Annomen.transcriptId': ['NR_024540.1'], 'Annomen.proteinId': [''], 'snp151.name': ['rs6682385'], 'snp151.subset': ['Common'], 'snp151.alleleN': ['2415'], 'snp151.alleleFreq': ['0.482228'], 'CADD.RawScore': ['-0.059553'], 'CADD.PHRED': ['1.269']}\n",
      "Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[26, 8], DP=34, GQ=99, PL=[142, 0, 791]))\n",
      "[Call(sample=D0_ACTB2, CallData(GT=0/1, AD=[26, 8], DP=34, GQ=99, PL=[142, 0, 791]))]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "vcfReader = vcf.Reader(open(vcfDir + vcfFile, 'r'))\n",
    "maxRecords = 10\n",
    "currentRecord = 1\n",
    "for record in vcfReader:\n",
    "    print(record)\n",
    "    print(record.INFO)\n",
    "    print(record.genotype('D0_ACTB2'))\n",
    "    print(record.samples)\n",
    "    print()\n",
    "    currentRecord += 1\n",
    "    if currentRecord > maxRecords:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "vcfReader = vcf.Reader(open(vcfDir + vcfFile, 'r'))\n",
    "maxRecords = 15\n",
    "currentRecord = 1\n",
    "variantList = []\n",
    "for record in vcfReader:\n",
    "    varFields={\n",
    "        'chr':record.CHROM,\n",
    "        'pos':record.POS,\n",
    "        'reference_allele':record.REF,\n",
    "    }\n",
    "    for s in record.samples:\n",
    "        call = record.genotype(s.sample)\n",
    "        callAlleles = call.data.GT.replace('|','/').split('/')\n",
    "        sampleFields={\n",
    "            'sample':s.sample,\n",
    "            'sample_depth':call.data.DP,\n",
    "            'quality':call.data.GQ\n",
    "        }\n",
    "        sampleFields.update(varFields)\n",
    "        allAlleles = [record.REF] + record.ALT\n",
    "        for alleleNum, depth in zip(callAlleles, call.data.AD):\n",
    "            sampleFields['alternate_allele'] = allAlleles[int(alleleNum)]\n",
    "            sampleFields['alternate_allele_depth'] = depth\n",
    "            variantList.append(sampleFields.copy())\n",
    "    currentRecord += 1\n",
    "    if currentRecord > maxRecords:\n",
    "        break\n",
    "variants = pd.DataFrame(variantList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 2,\n",
       "  'quality': 6,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14248,\n",
       "  'reference_allele': 'T',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 0},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 2,\n",
       "  'quality': 6,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14248,\n",
       "  'reference_allele': 'T',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 2},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 22,\n",
       "  'quality': 47,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14542,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': 'A',\n",
       "  'alternate_allele_depth': 18},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 22,\n",
       "  'quality': 47,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14542,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 4},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 34,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14574,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': 'A',\n",
       "  'alternate_allele_depth': 26},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 34,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14574,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 8},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 43,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14599,\n",
       "  'reference_allele': 'T',\n",
       "  'alternate_allele': 'T',\n",
       "  'alternate_allele_depth': 35},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 43,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14599,\n",
       "  'reference_allele': 'T',\n",
       "  'alternate_allele': A,\n",
       "  'alternate_allele_depth': 8},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 48,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14604,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': 'A',\n",
       "  'alternate_allele_depth': 40},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 48,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14604,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 8},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 51,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14610,\n",
       "  'reference_allele': 'T',\n",
       "  'alternate_allele': 'T',\n",
       "  'alternate_allele_depth': 42},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 51,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14610,\n",
       "  'reference_allele': 'T',\n",
       "  'alternate_allele': C,\n",
       "  'alternate_allele_depth': 9},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 103,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14653,\n",
       "  'reference_allele': 'C',\n",
       "  'alternate_allele': 'C',\n",
       "  'alternate_allele_depth': 68},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 103,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14653,\n",
       "  'reference_allele': 'C',\n",
       "  'alternate_allele': T,\n",
       "  'alternate_allele_depth': 35},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 120,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14677,\n",
       "  'reference_allele': 'G',\n",
       "  'alternate_allele': 'G',\n",
       "  'alternate_allele_depth': 96},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 120,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14677,\n",
       "  'reference_allele': 'G',\n",
       "  'alternate_allele': A,\n",
       "  'alternate_allele_depth': 24},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 41,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14907,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': 'A',\n",
       "  'alternate_allele_depth': 32},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 41,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14907,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 9},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 34,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14930,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': 'A',\n",
       "  'alternate_allele_depth': 26},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 34,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14930,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 8},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 34,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14932,\n",
       "  'reference_allele': 'G',\n",
       "  'alternate_allele': 'G',\n",
       "  'alternate_allele_depth': 27},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 34,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14932,\n",
       "  'reference_allele': 'G',\n",
       "  'alternate_allele': T,\n",
       "  'alternate_allele_depth': 7},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 35,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14937,\n",
       "  'reference_allele': 'T',\n",
       "  'alternate_allele': 'T',\n",
       "  'alternate_allele_depth': 28},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 35,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 14937,\n",
       "  'reference_allele': 'T',\n",
       "  'alternate_allele': C,\n",
       "  'alternate_allele_depth': 7},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 13,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 15118,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': 'A',\n",
       "  'alternate_allele_depth': 8},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 13,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 15118,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 5},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 2,\n",
       "  'quality': 6,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 15274,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 0},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 2,\n",
       "  'quality': 6,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 15274,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 2},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 15,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 15440,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': 'A',\n",
       "  'alternate_allele_depth': 11},\n",
       " {'sample': 'D0_ACTB2',\n",
       "  'sample_depth': 15,\n",
       "  'quality': 99,\n",
       "  'chr': 'chr1',\n",
       "  'pos': 15440,\n",
       "  'reference_allele': 'A',\n",
       "  'alternate_allele': G,\n",
       "  'alternate_allele_depth': 4}]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variantList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_depth</th>\n",
       "      <th>quality</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>alternate_allele</th>\n",
       "      <th>alternate_allele_depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14248</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14248</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>22</td>\n",
       "      <td>47</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14542</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>22</td>\n",
       "      <td>47</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14542</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>34</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14574</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>34</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14574</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>43</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14599</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>43</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14599</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>48</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14604</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>48</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14604</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>51</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14610</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>51</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14610</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>103</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14653</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>103</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14653</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>120</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14677</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>120</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14677</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>41</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14907</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>41</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14907</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>34</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14930</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>34</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14930</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>34</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14932</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>34</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14932</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>35</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14937</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>35</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14937</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>13</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15118</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>13</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15118</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15274</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15274</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>15</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15440</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>D0_ACTB2</td>\n",
       "      <td>15</td>\n",
       "      <td>99</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15440</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample  sample_depth  quality   chr    pos reference_allele  \\\n",
       "0   D0_ACTB2             2        6  chr1  14248                T   \n",
       "1   D0_ACTB2             2        6  chr1  14248                T   \n",
       "2   D0_ACTB2            22       47  chr1  14542                A   \n",
       "3   D0_ACTB2            22       47  chr1  14542                A   \n",
       "4   D0_ACTB2            34       99  chr1  14574                A   \n",
       "5   D0_ACTB2            34       99  chr1  14574                A   \n",
       "6   D0_ACTB2            43       99  chr1  14599                T   \n",
       "7   D0_ACTB2            43       99  chr1  14599                T   \n",
       "8   D0_ACTB2            48       99  chr1  14604                A   \n",
       "9   D0_ACTB2            48       99  chr1  14604                A   \n",
       "10  D0_ACTB2            51       99  chr1  14610                T   \n",
       "11  D0_ACTB2            51       99  chr1  14610                T   \n",
       "12  D0_ACTB2           103       99  chr1  14653                C   \n",
       "13  D0_ACTB2           103       99  chr1  14653                C   \n",
       "14  D0_ACTB2           120       99  chr1  14677                G   \n",
       "15  D0_ACTB2           120       99  chr1  14677                G   \n",
       "16  D0_ACTB2            41       99  chr1  14907                A   \n",
       "17  D0_ACTB2            41       99  chr1  14907                A   \n",
       "18  D0_ACTB2            34       99  chr1  14930                A   \n",
       "19  D0_ACTB2            34       99  chr1  14930                A   \n",
       "20  D0_ACTB2            34       99  chr1  14932                G   \n",
       "21  D0_ACTB2            34       99  chr1  14932                G   \n",
       "22  D0_ACTB2            35       99  chr1  14937                T   \n",
       "23  D0_ACTB2            35       99  chr1  14937                T   \n",
       "24  D0_ACTB2            13       99  chr1  15118                A   \n",
       "25  D0_ACTB2            13       99  chr1  15118                A   \n",
       "26  D0_ACTB2             2        6  chr1  15274                A   \n",
       "27  D0_ACTB2             2        6  chr1  15274                A   \n",
       "28  D0_ACTB2            15       99  chr1  15440                A   \n",
       "29  D0_ACTB2            15       99  chr1  15440                A   \n",
       "\n",
       "   alternate_allele  alternate_allele_depth  \n",
       "0                 G                       0  \n",
       "1                 G                       2  \n",
       "2                 A                      18  \n",
       "3                 G                       4  \n",
       "4                 A                      26  \n",
       "5                 G                       8  \n",
       "6                 T                      35  \n",
       "7                 A                       8  \n",
       "8                 A                      40  \n",
       "9                 G                       8  \n",
       "10                T                      42  \n",
       "11                C                       9  \n",
       "12                C                      68  \n",
       "13                T                      35  \n",
       "14                G                      96  \n",
       "15                A                      24  \n",
       "16                A                      32  \n",
       "17                G                       9  \n",
       "18                A                      26  \n",
       "19                G                       8  \n",
       "20                G                      27  \n",
       "21                T                       7  \n",
       "22                T                      28  \n",
       "23                C                       7  \n",
       "24                A                       8  \n",
       "25                G                       5  \n",
       "26                G                       0  \n",
       "27                G                       2  \n",
       "28                A                      11  \n",
       "29                G                       4  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "not enough values to unpack (expected 2, got 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-87-8946f196404d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0md\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m'a'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'b'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0md\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: not enough values to unpack (expected 2, got 1)"
     ]
    }
   ],
   "source": [
    "d={'a':1,'b':2}\n",
    "for k,v in d:\n",
    "    print(k,v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting hgvs\n",
      "  Downloading hgvs-1.5.1-py2.py3-none-any.whl (98 kB)\n",
      "\u001b[K     |################################| 98 kB 4.9 MB/s eta 0:00:011\n",
      "\u001b[?25hCollecting psycopg2-binary\n",
      "  Downloading psycopg2_binary-2.8.6-cp36-cp36m-manylinux1_x86_64.whl (3.0 MB)\n",
      "\u001b[K     |################################| 3.0 MB 12.8 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: ipython in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from hgvs) (7.16.1)\n",
      "Collecting configparser>=3.3.0\n",
      "  Downloading configparser-5.0.1-py3-none-any.whl (22 kB)\n",
      "Requirement already satisfied: six in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from hgvs) (1.15.0)\n",
      "Collecting parsley\n",
      "  Downloading Parsley-1.3-py2.py3-none-any.whl (88 kB)\n",
      "\u001b[K     |################################| 88 kB 8.1 MB/s  eta 0:00:01\n",
      "\u001b[?25hCollecting biocommons.seqrepo<1.0\n",
      "  Downloading biocommons.seqrepo-0.6.3-py2.py3-none-any.whl (31 kB)\n",
      "Requirement already satisfied: attrs>=17.4.0 in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from hgvs) (20.3.0)\n",
      "Collecting bioutils<1.0,>=0.4.0\n",
      "  Downloading bioutils-0.5.2.post3-py2.py3-none-any.whl (256 kB)\n",
      "\u001b[K     |################################| 256 kB 11.6 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from ipython->hgvs) (3.0.8)\n",
      "Requirement already satisfied: setuptools>=18.5 in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from ipython->hgvs) (49.2.0.post20200712)\n",
      "Requirement already satisfied: pygments in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from ipython->hgvs) (2.7.2)\n",
      "Requirement already satisfied: pexpect; sys_platform != \"win32\" in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from ipython->hgvs) (4.8.0)\n",
      "Requirement already satisfied: jedi>=0.10 in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from ipython->hgvs) (0.17.2)\n",
      "Requirement already satisfied: traitlets>=4.2 in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from ipython->hgvs) (4.3.3)\n",
      "Requirement already satisfied: decorator in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from ipython->hgvs) (4.4.2)\n",
      "Requirement already satisfied: pickleshare in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from ipython->hgvs) (0.7.5)\n",
      "Requirement already satisfied: backcall in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from ipython->hgvs) (0.2.0)\n",
      "Collecting requests\n",
      "  Downloading requests-2.25.1-py2.py3-none-any.whl (61 kB)\n",
      "\u001b[K     |################################| 61 kB 8.1 MB/s  eta 0:00:01\n",
      "\u001b[?25hCollecting yoyo-migrations\n",
      "  Downloading yoyo_migrations-7.2.1-py3-none-any.whl (36 kB)\n",
      "Requirement already satisfied: tqdm in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from biocommons.seqrepo<1.0->hgvs) (4.48.0)\n",
      "Collecting requests-html\n",
      "  Downloading requests_html-0.10.0-py3-none-any.whl (13 kB)\n",
      "Collecting coloredlogs\n",
      "  Downloading coloredlogs-15.0-py2.py3-none-any.whl (45 kB)\n",
      "\u001b[K     |################################| 45 kB 5.2 MB/s  eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: pysam in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from biocommons.seqrepo<1.0->hgvs) (0.16.0.1)\n",
      "Requirement already satisfied: wcwidth in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython->hgvs) (0.2.5)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from pexpect; sys_platform != \"win32\"->ipython->hgvs) (0.6.0)\n",
      "Requirement already satisfied: parso<0.8.0,>=0.7.0 in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from jedi>=0.10->ipython->hgvs) (0.7.1)\n",
      "Requirement already satisfied: ipython-genutils in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from traitlets>=4.2->ipython->hgvs) (0.2.0)\n",
      "Collecting chardet<5,>=3.0.2\n",
      "  Downloading chardet-4.0.0-py2.py3-none-any.whl (178 kB)\n",
      "\u001b[K     |################################| 178 kB 11.5 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: certifi>=2017.4.17 in /opt/miniconda/envs/gatk/lib/python3.6/site-packages (from requests->biocommons.seqrepo<1.0->hgvs) (2020.6.20)\n",
      "Collecting idna<3,>=2.5\n",
      "  Downloading idna-2.10-py2.py3-none-any.whl (58 kB)\n",
      "\u001b[K     |################################| 58 kB 7.4 MB/s  eta 0:00:01\n",
      "\u001b[?25hCollecting urllib3<1.27,>=1.21.1\n",
      "  Downloading urllib3-1.26.2-py2.py3-none-any.whl (136 kB)\n",
      "\u001b[K     |################################| 136 kB 12.0 MB/s eta 0:00:01\n",
      "\u001b[?25hCollecting tabulate\n",
      "  Downloading tabulate-0.8.7-py3-none-any.whl (24 kB)\n",
      "Collecting sqlparse\n",
      "  Downloading sqlparse-0.4.1-py3-none-any.whl (42 kB)\n",
      "\u001b[K     |################################| 42 kB 2.0 MB/s  eta 0:00:01\n",
      "\u001b[?25hCollecting fake-useragent\n",
      "  Downloading fake-useragent-0.1.11.tar.gz (13 kB)\n",
      "Collecting bs4\n",
      "  Downloading bs4-0.0.1.tar.gz (1.1 kB)\n",
      "Collecting pyquery\n",
      "  Downloading pyquery-1.4.3-py3-none-any.whl (22 kB)\n",
      "Collecting pyppeteer>=0.0.14\n",
      "  Downloading pyppeteer-0.2.4-py3-none-any.whl (88 kB)\n",
      "\u001b[K     |################################| 88 kB 7.8 MB/s  eta 0:00:01\n",
      "\u001b[?25hCollecting w3lib\n",
      "  Downloading w3lib-1.22.0-py2.py3-none-any.whl (20 kB)\n",
      "Collecting parse\n",
      "  Downloading parse-1.18.0.tar.gz (30 kB)\n",
      "Collecting humanfriendly>=9.1\n",
      "  Downloading humanfriendly-9.1-py2.py3-none-any.whl (86 kB)\n",
      "\u001b[K     |################################| 86 kB 7.0 MB/s  eta 0:00:01\n",
      "\u001b[?25hCollecting beautifulsoup4\n",
      "  Downloading beautifulsoup4-4.9.3-py3-none-any.whl (115 kB)\n",
      "\u001b[K     |################################| 115 kB 12.0 MB/s eta 0:00:01\n",
      "\u001b[?25hCollecting lxml>=2.1\n",
      "  Downloading lxml-4.6.2-cp36-cp36m-manylinux1_x86_64.whl (5.5 MB)\n",
      "\u001b[K     |################################| 5.5 MB 11.2 MB/s eta 0:00:01\n",
      "\u001b[?25hCollecting cssselect>0.7.9\n",
      "  Downloading cssselect-1.1.0-py2.py3-none-any.whl (16 kB)\n",
      "Collecting appdirs<2.0.0,>=1.4.3\n",
      "  Downloading appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)\n",
      "Collecting pyee<9.0.0,>=8.1.0\n",
      "  Downloading pyee-8.1.0-py2.py3-none-any.whl (12 kB)\n",
      "Collecting websockets<9.0,>=8.1\n",
      "  Downloading websockets-8.1-cp36-cp36m-manylinux2010_x86_64.whl (78 kB)\n",
      "\u001b[K     |################################| 78 kB 7.7 MB/s  eta 0:00:01\n",
      "\u001b[?25hCollecting soupsieve>1.2; python_version >= \"3.0\"\n",
      "  Downloading soupsieve-2.1-py3-none-any.whl (32 kB)\n",
      "Building wheels for collected packages: fake-useragent, bs4, parse\n",
      "  Building wheel for fake-useragent (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for fake-useragent: filename=fake_useragent-0.1.11-py3-none-any.whl size=13487 sha256=ccace3685d117096a58391dae1d37d7a766a461f6ce698b138c2a6e149582f29\n",
      "  Stored in directory: /root/.cache/pip/wheels/12/8a/a1/0ab79c1c2c58aa7647d0dea17f2af2358b65e1eecac8094b5e\n",
      "  Building wheel for bs4 (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for bs4: filename=bs4-0.0.1-py3-none-any.whl size=1272 sha256=3bb36410bd1c2c4362fb1eec5056d8e11945339fc5840c842100811a451f4a19\n",
      "  Stored in directory: /root/.cache/pip/wheels/19/f5/6d/a97dd4f22376d4472d5f4c76c7646876052ff3166b3cf71050\n",
      "  Building wheel for parse (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for parse: filename=parse-1.18.0-py3-none-any.whl size=24133 sha256=53a4119cb1114ef1a71c043d040e069e839225a85fb9810970544d64d69d6bd5\n",
      "  Stored in directory: /root/.cache/pip/wheels/85/db/d9/2416d9d65785454955833a9272db1df200ee64f2269cf6e282\n",
      "Successfully built fake-useragent bs4 parse\n",
      "Installing collected packages: psycopg2-binary, configparser, parsley, chardet, idna, urllib3, requests, tabulate, sqlparse, yoyo-migrations, bioutils, fake-useragent, soupsieve, beautifulsoup4, bs4, lxml, cssselect, pyquery, appdirs, pyee, websockets, pyppeteer, w3lib, parse, requests-html, humanfriendly, coloredlogs, biocommons.seqrepo, hgvs\n",
      "Successfully installed appdirs-1.4.4 beautifulsoup4-4.9.3 biocommons.seqrepo-0.6.3 bioutils-0.5.2.post3 bs4-0.0.1 chardet-4.0.0 coloredlogs-15.0 configparser-5.0.1 cssselect-1.1.0 fake-useragent-0.1.11 hgvs-1.5.1 humanfriendly-9.1 idna-2.10 lxml-4.6.2 parse-1.18.0 parsley-1.3 psycopg2-binary-2.8.6 pyee-8.1.0 pyppeteer-0.2.4 pyquery-1.4.3 requests-2.25.1 requests-html-0.10.0 soupsieve-2.1 sqlparse-0.4.1 tabulate-0.8.7 urllib3-1.26.2 w3lib-1.22.0 websockets-8.1 yoyo-migrations-7.2.1\n",
      "\u001b[33mWARNING: You are using pip version 20.2.4; however, version 20.3.3 is available.\n",
      "You should consider upgrading via the '/opt/miniconda/envs/gatk/bin/python3.6 -m pip install --upgrade pip' command.\u001b[0m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install hgvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hgvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SequenceVariant(ac=NC_000007.13, type=g, posedit=36561662C>T, gene=None)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import hgvs.parser\n",
    "hgvs_g = 'NC_000007.13:g.36561662C>T'\n",
    "hgvs_c = 'NM_001637.3:c.1582G>A'\n",
    "hp = hgvs.parser.Parser()\n",
    "var_g = hp.parse_hgvs_variant(hgvs_g)\n",
    "var_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SimplePosition(base=36561662, uncertain=False)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_g.posedit.pos.start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NC_000007.13:g.36561662C>T'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(var_g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SequenceVariant(ac=NM_001637.3, type=c, posedit=1582G>A, gene=None)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import hgvs.dataproviders.uta\n",
    "import hgvs.assemblymapper\n",
    "\n",
    "# initialize the mapper for GRCh37 with splign-based alignments\n",
    "hdp = hgvs.dataproviders.uta.connect()\n",
    "am = hgvs.assemblymapper.AssemblyMapper(hdp,\n",
    "         assembly_name='GRCh37', alt_aln_method='splign',\n",
    "         replace_reference=True)\n",
    "var_c = am.g_to_c(var_g, 'NM_001637.3')\n",
    "var_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record(CHROM=chr1, POS=14248, REF=T, ALT=[G])\n",
      "G|downstream_gene_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene||n.*156A>C|||||156|\n",
      "G|downstream_gene_variant|MODIFIER|MIR6859-1|ENSG00000278267|transcript|ENST00000619216.1|miRNA||n.*3121A>C|||||3121|\n",
      "G|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000450305.2|transcribed_unprocessed_pseudogene||n.*578T>G|||||578|\n",
      "G|non_coding_transcript_exon_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript|3/3|n.1496T>G||||||\n",
      "\n",
      "Record(CHROM=chr1, POS=14542, REF=A, ALT=[G])\n",
      "G|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.*133A>G|||||133|\n",
      "G|downstream_gene_variant|MODIFIER|MIR6859-1|ENSG00000278267|transcript|ENST00000619216.1|miRNA||n.*2827T>C|||||2827|\n",
      "G|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000450305.2|transcribed_unprocessed_pseudogene||n.*872A>G|||||872|\n",
      "G|intron_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene|10/10|n.1254-41T>C||||||\n",
      "\n",
      "Record(CHROM=chr1, POS=14574, REF=A, ALT=[G])\n",
      "G|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.*165A>G|||||165|\n",
      "G|downstream_gene_variant|MODIFIER|MIR6859-1|ENSG00000278267|transcript|ENST00000619216.1|miRNA||n.*2795T>C|||||2795|\n",
      "G|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000450305.2|transcribed_unprocessed_pseudogene||n.*904A>G|||||904|\n",
      "G|intron_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene|10/10|n.1254-73T>C||||||\n",
      "\n",
      "Record(CHROM=chr1, POS=14599, REF=T, ALT=[A])\n",
      "A|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.*190T>A|||||190|\n",
      "A|downstream_gene_variant|MODIFIER|MIR6859-1|ENSG00000278267|transcript|ENST00000619216.1|miRNA||n.*2770A>T|||||2770|\n",
      "A|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000450305.2|transcribed_unprocessed_pseudogene||n.*929T>A|||||929|\n",
      "A|intron_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene|10/10|n.1254-98A>T||||||\n",
      "\n",
      "Record(CHROM=chr1, POS=14604, REF=A, ALT=[G])\n",
      "G|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.*195A>G|||||195|\n",
      "G|downstream_gene_variant|MODIFIER|MIR6859-1|ENSG00000278267|transcript|ENST00000619216.1|miRNA||n.*2765T>C|||||2765|\n",
      "G|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000450305.2|transcribed_unprocessed_pseudogene||n.*934A>G|||||934|\n",
      "G|intron_variant|MODIFIER|WASH7P|ENSG00000227232|transcript|ENST00000488147.1|unprocessed_pseudogene|10/10|n.1254-103T>C||||||\n",
      "\n"
     ]
    }
   ],
   "source": [
    "vcfDir='/data/nijhawanlab/indisulam/'\n",
    "vcfFile='D0_ACTB2.filtered.snpEff.vcf'\n",
    "vcfReader = vcf.Reader(open(vcfDir + vcfFile, 'r'))\n",
    "maxRecords = 5\n",
    "currentRecord = 1\n",
    "for record in vcfReader:\n",
    "    print(record)\n",
    "    annotations = record.INFO['ANN']\n",
    "    for a in annotations:\n",
    "        print(a)\n",
    "    print()\n",
    "    currentRecord += 1\n",
    "    if currentRecord > maxRecords:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "a='G|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.*195A>G|||||195|'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'G|downstream_gene_variant|MODIFIER|DDX11L1|ENSG00000223972|transcript|ENST00000456328.2|processed_transcript||n.*195A>G|||||195|'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['G',\n",
       " 'downstream_gene_variant',\n",
       " 'MODIFIER',\n",
       " 'DDX11L1',\n",
       " 'ENSG00000223972',\n",
       " 'transcript',\n",
       " 'ENST00000456328.2',\n",
       " 'processed_transcript',\n",
       " '',\n",
       " 'n.*195A>G',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '195',\n",
       " '']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.split('|')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "ANNFields = ['allele',\n",
    "             'annotation',\n",
    "             'putative_impact',\n",
    "             'gene_name',\n",
    "             'gene_id',\n",
    "             'feature_type',\n",
    "             'feature_id',\n",
    "             'transcript_biotype',\n",
    "             'rank',\n",
    "             'HGVS.c',\n",
    "             'GHVS.p',\n",
    "             'cDNA_position',\n",
    "             'cds_position',\n",
    "             'protein_position',\n",
    "             'distance_to_feature',\n",
    "             'messages'\n",
    "             ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'allele': 'G',\n",
       " 'annotation': 'downstream_gene_variant',\n",
       " 'putative_impact': 'MODIFIER',\n",
       " 'gene_name': 'DDX11L1',\n",
       " 'gene_id': 'ENSG00000223972',\n",
       " 'feature_type': 'transcript',\n",
       " 'feature_id': 'ENST00000456328.2',\n",
       " 'transcript_biotype': 'processed_transcript',\n",
       " 'rank': '',\n",
       " 'HGVS.c': 'n.*195A>G',\n",
       " 'GHVS.p': '',\n",
       " 'cDNA_position': '',\n",
       " 'cds_position': '',\n",
       " 'protein_position': '',\n",
       " 'distance_to_feature': '195',\n",
       " 'messages': ''}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d={k:v for k,v in zip(ANNFields, a.split('|'))}\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['allele',\n",
       " 'annotation',\n",
       " 'putative_impact',\n",
       " 'gene_name',\n",
       " 'gene_id',\n",
       " 'feature_type',\n",
       " 'feature_id',\n",
       " 'transcript_biotype',\n",
       " 'rank',\n",
       " 'HGVS.c',\n",
       " 'GHVS.p',\n",
       " 'cDNA_position',\n",
       " 'cds_position',\n",
       " 'protein_position',\n",
       " 'distance_to_feature',\n",
       " 'messages',\n",
       " ['A', 'B', 'C'],\n",
       " ['A', 'B', 'C'],\n",
       " 'A',\n",
       " 'B',\n",
       " 'C']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = ['A','B','C']\n",
    "ANNFields.extend(l)\n",
    "ANNFields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
